README.md
# 5G00EV17-3001 Unit testing

An example project for showcasing unit testing.

|   Category   |  Badges  |
|:------------:|---|
| **PyPI**     | ![Python versions](https://img.shields.io/pypi/pyversions/unit-testing?logo=python) ![PyPI - Implementation](https://img.shields.io/pypi/implementation/unit-testing) |
| **Tests**    | [![codecov](https://codecov.io/gh/Diapolo10/5G00EV17-3001_unit-testing/branch/main/graph/badge.svg?token=FpaCuVjOAB)](https://codecov.io/gh/Diapolo10/5G00EV17-3001_unit-testing) ![Unit tests](https://github.com/diapolo10/5G00EV17-3001_unit-testing/workflows/Unit%20tests/badge.svg) ![Ruff](https://github.com/diapolo10/5G00EV17-3001_unit-testing/workflows/Ruff/badge.svg) |
| **Activity** | ![GitHub contributors](https://img.shields.io/github/contributors/diapolo10/5G00EV17-3001_unit-testing) ![Last commit](https://img.shields.io/github/last-commit/diapolo10/5G00EV17-3001_unit-testing?logo=github)
| **QA**       | [![CodeFactor](https://www.codefactor.io/repository/github/diapolo10/5G00EV17-3001_unit-testing/badge?logo=codefactor)](https://www.codefactor.io/repository/github/diapolo10/5G00EV17-3001_unit-testing) [![Rating](https://img.shields.io/librariesio/sourcerank/pypi/unit-testing)](https://libraries.io/github/Diapolo10/unit-testing/sourcerank) |
| **Other**    | [![License](https://img.shields.io/github/license/diapolo10/5G00EV17-3001_unit-testing)](https://opensource.org/licenses/MIT) ![Repository size](https://img.shields.io/github/repo-size/diapolo10/5G00EV17-3001_unit-testing?logo=github) ![Code size](https://img.shields.io/github/languages/code-size/diapolo10/5G00EV17-3001_unit-testing?logo=github) ![Lines of code](https://img.shields.io/tokei/lines/github/diapolo10/5G00EV17-3001_unit-testing?logo=github) |

-------------------------------------------------------------------------------

## Description

This project showcases how unit testing works by implementing a very simple
Python library, and adding unit tests to it. The project uses Pytest.

## Getting Started

### Dependencies

The main project has no library dependencies, but the actual unit testing
part relies on several packages listed in
[`pyproject.toml`][pyproject.toml]. In general, you'll need:

- Python 3.11 or newer
- Poetry

The project is automatically tested on the latest versions of Windows,
Mac OS, and Ubuntu, and it has also been tested on both CPython
and PyPy. Using other implementations or operating systems
may work, but is not guaranteed.

### Installation

Please see the documentation [here][installation].

### Running unit tests

Please see the documentation [here][running unit tests].

## Version history

The project's changelog can be found [here][changelog].

## License

This project is licensed under the MIT license - see the [`LICENSE`][license]-file for details.

## Acknowledgements

Inspiration, code snippets, debugging help, etc.

- My fellow team members

[pyproject.toml]: ./pyproject.toml
[installation]: ./docs/installation.md
[running unit tests]: ./docs/running_unit_tests.md
[changelog]: ./CHANGELOG.md
[license]: ./LICENSE


CHANGELOG.md

# 5G00EV17-3001 Unit testing Changelog

All notable changes to this project will be documented in this file.

The format is based on [CHANGELOG.md][CHANGELOG.md]
and this project adheres to [Semantic Versioning][Semantic Versioning].

<!-- 
TEMPLATE

## [major.minor.patch] - yyyy-mm-dd

A message that notes the main changes in the update.

### Added

### Changed

### Deprecated

### Fixed

### Removed

### Security

_______________________________________________________________________________
 
 -->

<!--
EXAMPLE

## [0.2.0] - 2021-06-02

Lorem Ipsum dolor sit amet.

### Added

- Cat pictures hidden in the library
- Added beeswax to the gears

### Changed

- Updated localisation files

-->

<!--
_______________________________________________________________________________

## [1.0.4] - 2023-11-02

Updated dependencies, and updated Tox configuration to fix coverage
report generation.

### Changed

- Updated dependencies
- Updated localisation files

### Fixed

- Coverage report generation via Tox was broken, this update
  fixes that by changing how Tox is configured.

-->

_______________________________________________________________________________

## [1.0.4] - 2023-11-02

Updated dependencies, and updated Tox configuration to fix coverage
report generation.

### Changed

- Updated dependencies
- Updated localisation files

### Fixed

- Coverage report generation via Tox was broken, this update
  fixes that by changing how Tox is configured.

_______________________________________________________________________________

## [1.0.3] - 2023-10-03

Updated dependencies, moved source code to `src`, made Ruff rules stricter,
and fixed outdated `Makefile`.

### Added

- Added new rules for Ruff
- Added hash support for the `Date` class

### Changed

- Moved the source code files under `src` to better match accepted best
  practices
- Updated outdated links in `CHANGELOG.md`
- Updated dependencies
- Updated localisation files

### Fixed

- Fixed `Makefile`'s linting (it was still using `pyproject-flake8`
  instead of Ruff)
- Various linter fixes
- `README.md` had an old version number for Python, this is now fixed

_______________________________________________________________________________

## [1.0.2] - 2023-02-24

This update adds a missing `py.typed` file, and updates dependencies.

### Added

- `py.typed`

### Changed

- Fixed some wording in `README.md`
- Updated dependencies
- Updated localisation files

_______________________________________________________________________________

## [1.0.1] - 2022-04-12

This update adds some additional test cases for corner cases, improves
type hinting, and adds missing docstrings

### Added

- More docstrings
- More test cases
- Documentation is now linked in `README.md`

### Changed

- Improved type hinting
- Updated localisation files

_______________________________________________________________________________

## [1.0.0] - 2022-03-30

This is the initial version of the project.

### Added

- Example project featuring a Date class
- Unit tests
- Coverage reports
- Documentation for installation and unit tests

### Changed

- Updated `README.md` to contain more useful information
- Updated localisation files

### Fixed

- Linter errors

### Removed

- Dread

### Security

- Project has Snyk scanning enabled

[CHANGELOG.md]: https://keepachangelog.com/en/1.1.0/
[Semantic Versioning]: http://semver.org/

<!-- markdownlint-configure-file {
    "MD022": false,
    "MD024": false,
    "MD030": false,
    "MD032": false
} -->
<!--
    MD022: Blanks around headings
    MD024: No duplicate headings
    MD030: Spaces after list markers
    MD032: Blanks around lists
-->


tests/conftest.py
"""Fixtures for use in test cases."""

import pytest

from unit_testing import date


@pytest.fixture()
def default_date() -> date.Date:
    """Create a default date.Date object for testing."""
    return date.Date()


tests/test_date.py
"""Unit tests for the date.Date class."""

from collections.abc import Generator

import pytest

from unit_testing.date import (
    DEFAULT_DAY,
    DEFAULT_MONTH,
    DEFAULT_YEAR,
    Date,
    Months,
    week_later,
)


def test_date_default_attributes(default_date: Date) -> None:
    """Tests the default state of the date object."""
    assert default_date.day == DEFAULT_DAY
    assert default_date.month == DEFAULT_MONTH
    assert default_date.year == DEFAULT_YEAR


def test_date_default_methods(default_date: Date) -> None:
    """Tests the methods on a default date object."""
    assert default_date.leap_year() is False

    default_date.next_day()

    assert default_date.day == DEFAULT_DAY+1
    assert default_date.month == DEFAULT_MONTH
    assert default_date.year == DEFAULT_YEAR


def test_date_next_day(default_date: Date) -> None:
    """Tests setting the 'previous'-parameter for Date.next_day."""
    date = Date(30, Months.MARCH, 2022)
    default_date.next_day(previous=date)

    assert default_date.day == 31  # noqa: PLR2004
    assert default_date.month == Months.MARCH
    assert default_date.year == 2022  # noqa: PLR2004


def test_date_year_change() -> None:
    """Tests if year change is correctly implemented."""
    date = Date(31, Months.DECEMBER, 1999)
    date.next_day()

    assert date.day == 1
    assert date.month == 1
    assert date.year == 2000  # noqa: PLR2004
    assert date.leap_year() is True


def test_date_month_change() -> None:
    """Tests if month change is correctly implemented."""
    date = Date(28, Months.FEBRUARY, 2001)
    date.next_day()

    assert date.day == 1
    assert date.month == 3  # noqa: PLR2004
    assert date.year == 2001  # noqa: PLR2004


def test_date_day_error() -> None:
    """Tests that invalid days raise an error."""
    with pytest.raises(ValueError, match=" is not a valid day count for the given month"):
        Date(31, Months.FEBRUARY, 2020)


def test_date_month_error() -> None:
    """Tests that invalid months raise an error."""
    with pytest.raises(ValueError, match=" is not a valid month"):
        Date(7, 21, 1998)


def test_date_print(default_date: Date, capsys: Generator[pytest.CaptureFixture[str], None, None]) -> None:
    """'Tests' the printing functionality."""
    default_date.print()
    captured = capsys.readouterr()  # type: ignore[attr-defined]
    assert captured.out == f'{DEFAULT_DAY}/{DEFAULT_MONTH}/{DEFAULT_YEAR}\n'


def test_date_comparisons() -> None:
    """Tests comparison operators for Date objects."""
    date_1 = Date(1, 1, 2000)
    date_2 = Date(6, 1, 2000)
    date_3 = Date(24, 12, 1999)
    date_4 = Date(2, 2, 2000)
    date_5 = Date(1, 1, 2000)

    assert date_1 < date_2
    assert date_2 > date_1
    assert date_1 <= date_2
    assert date_2 >= date_1

    assert date_1 > date_3
    assert date_3 < date_1
    assert date_1 >= date_3
    assert date_3 <= date_1

    assert date_1 < date_4
    assert date_4 > date_1
    assert date_1 <= date_4
    assert date_4 >= date_1

    assert date_1 == date_5
    assert date_1 <= date_5
    assert date_1 >= date_5

    assert (date_1 < None) is False
    assert (date_1 > None) is False
    assert (date_1 <= None) is False
    assert (date_1 >= None) is False
    assert (date_1 == 'foo') is False


def test_date_leap_year() -> None:
    """Tests that leap years are correctly implemented."""
    date_1 = Date(1, 1, 2000)
    date_2 = Date(1, 1, 2100)
    date_3 = Date(1, 1, 2004)
    date_4 = Date(1, 1, 2001)

    assert date_1.leap_year() is True
    assert date_2.leap_year() is False
    assert date_3.leap_year() is True
    assert date_4.leap_year() is False


def test_date_days_between() -> None:
    """Tests calculating difference in days betweeen dates."""
    date_1 = Date(24, 1, 1998)
    date_2 = Date(31, 1, 1998)
    date_3 = Date(5, 2, 1998)
    date_4 = Date(24, 1, 2005)

    assert date_1.days_between(date_2) == 7  # noqa: PLR2004
    assert date_1.days_between(date_3) == 12  # noqa: PLR2004
    assert date_1.days_between(date_4) == 2557  # noqa: PLR2004
    assert date_4.days_between(date_1) == 2557  # noqa: PLR2004


def test_date_days_since_new_year() -> None:
    """Tests calculating number of days since the last new year."""
    date_1 = Date(31, 1, 2022)
    date_2 = Date(31, 12, 2022)

    assert date_1.days_since_new_year() == 30  # noqa: PLR2004
    assert date_2.days_since_new_year() == 364  # noqa: PLR2004


def test_week_later(default_date: Date) -> None:
    """Tests calculating the data a week later."""
    new_date = week_later(default_date)

    assert new_date.day == DEFAULT_DAY+7
    assert new_date.month == DEFAULT_MONTH
    assert new_date.year == DEFAULT_YEAR


tests/test_default.py
"""Stub unit test file"""

import pytest

from unit_testing import date

@pytest.fixture
def default_date():
    return date.Date()


def test_default_attributes(default_date):
    assert default_date.day == 1
    assert default_date.month == 1
    assert default_date.year == 1900


def test_default_methods(default_date):
    assert default_date.leap_year() == False
    
    default_date.next_day()

    assert default_date.day == 2
    assert default_date.month == 1
    assert default_date.year == 1900


def test_year_change():
    d = date.Date(31, 12, 1999)
    d.next_day()

    assert d.day == 1
    assert d.month == 1
    assert d.year == 2000
    assert d.leap_year() == True


def test_month_change():
    d = date.Date(28, 2, 2001)
    d.next_day()

    assert d.day == 1
    assert d.month == 3
    assert d.year == 2001


def test_date_error():
    with pytest.raises(ValueError):
        d = date.Date(31, 2, 2020)


def test_month_error():
    with pytest.raises(ValueError):
        d = date.Date(7, 21, 1998)


src/unit_testing/date.py
"""Contains the implementation of the Date class."""

from __future__ import annotations

from enum import IntEnum

DEFAULT_DAY = 1
DEFAULT_MONTH = 1
DEFAULT_YEAR = 1900
DAYS_IN_FEBRUARY = 28
DAYS_IN_OTHER_MONTHS = 30
MAX_DAYS_IN_MONTH = 31


class Months(IntEnum):
    """Enum for months."""

    JANUARY   = 1
    FEBRUARY  = 2
    MARCH     = 3
    APRIL     = 4
    MAY       = 5
    JUNE      = 6
    JULY      = 7
    AUGUST    = 8
    SEPTEMBER = 9
    OCTOBER   = 10
    NOVEMBER  = 11
    DECEMBER  = 12


class Date:
    """Date implementation."""

    def __init__(self: Date, day: int | None = None, month: int | None = None, year: int | None = None) -> None:
        """Initialise Date object."""
        self.year: int = year if year is not None else DEFAULT_YEAR
        self.month: int = month if month is not None else DEFAULT_MONTH
        self.day: int = day if day is not None else DEFAULT_DAY

    @property
    def day(self: Date) -> int:
        """Getter for days."""
        return self._day

    @day.setter
    def day(self: Date, other: int) -> None:
        """Setter for days."""
        if other > self.days_in_month():
            msg = f"{other} is not a valid day count for the given month"
            raise ValueError(msg)
        self._day = other

    @property
    def month(self: Date) -> int:
        """Getter for months."""
        return self._month

    @month.setter
    def month(self: Date, other: int) -> None:
        """Setter for months."""
        if not Months.JANUARY <= other <= Months.DECEMBER:
            msg = f"{other} is not a valid month"
            raise ValueError(msg)
        self._month = other

    @property
    def year(self: Date) -> int:
        """Getter for years."""
        return self._year

    @year.setter
    def year(self: Date, other: int) -> None:
        """Setter for years."""
        self._year = other

    def __str__(self: Date) -> str:
        """Date string representation."""
        return f"{self.day}/{self.month}/{self.year}"

    def __gt__(self: Date, other: object) -> bool:
        """Implement greater than for Date objects."""
        if isinstance(other, Date):
            if self.year > other.year:
                return True
            if self.year == other.year and self.month > other.month:
                return True
            if self.year == other.year and self.month == other.month and self.day > other.day:
                return True
        return False

    def __ge__(self: Date, other: object) -> bool:
        """Implement greater than or equal for Date objects."""
        return self > other or self == other

    def __le__(self: Date, other: object) -> bool:
        """Implement less than or equal for Date objects."""
        return self < other or self == other

    def __lt__(self: Date, other: object) -> bool:
        """Implement less than for Date objects."""
        if isinstance(other, Date):
            return not self >= other
        return False

    def __eq__(self: Date, other: object) -> bool:
        """Check for date object equality."""
        if isinstance(other, Date):
            return hash(self) == hash(other)

        return False

    def __hash__(self: Date) -> int:
        """Hash the Date object."""
        return hash((self.year, self.month, self.day))

    def days_in_month(self: Date) -> int:
        """Count the days in a given month."""
        max_days = MAX_DAYS_IN_MONTH
        if self.month == Months.FEBRUARY:
            max_days = DAYS_IN_FEBRUARY
            if self.leap_year():
                max_days += 1
        elif self.month in {Months.APRIL, Months.JUNE, Months.SEPTEMBER, Months.NOVEMBER}:
            max_days = DAYS_IN_OTHER_MONTHS
        return max_days

    def next_day(self: Date, previous: Date | None = None) -> None:
        """Move to the next day."""
        if previous is not None:
            self.year = previous.year
            self.month = previous.month
            self.day = previous.day

        if self.day + 1 > self.days_in_month():
            if self.month + 1 > Months.DECEMBER:
                self.year += 1
                self.month = 1
            else:
                self.month += 1
            self.day = 1
        else:
            self.day += 1

    def print(self: Date) -> None:  # noqa: A003
        """Print the current day."""
        print(self)  # noqa: T201

    def leap_year(self: Date) -> bool:
        """Check if the current year is a leap year."""
        if self.year % 400 == 0:
            return True
        if self.year % 100 == 0:
            return False

        return self.year % 4 == 0

    def days_between(self: Date, other: Date) -> int:
        """Return the number of days between two given date objects."""
        days = 0

        destination = self
        start = Date(other.day, other.month, other.year)
        if self < other:
            destination = other
            start = Date(self.day, self.month, self.year)

        while start < destination:
            start.next_day()
            days += 1

        return days

    def days_since_new_year(self: Date) -> int:
        """Return the number of days since new year."""
        new_year = Date(1, 1, self.year)
        return self.days_between(new_year)


def week_later(date: Date) -> Date:
    """Return a new Date object a week later from the given date."""
    new_date = Date(date.day, date.month, date.year)
    for _ in range(7):
        new_date.next_day()
    return new_date
